File: src/shader.cpp
#include "shader.h"

void Shader::shaderErrorCheck(GLuint* shader){
	int success;
	char infoLog[512];
	glGetShaderiv(*shader, GL_COMPILE_STATUS, &success);
	if(!success){
		glGetShaderInfoLog(*shader, 512, NULL, infoLog);
		std::cout << "Error compiling shader: " << infoLog << std::endl;
	}
}
void Shader::programErrorCheck(GLuint* program){
	int success;
	char infoLog[512];
	glGetShaderiv(*program, GL_LINK_STATUS, &success);
	if(!success){
		glGetProgramInfoLog(*program, 512, NULL, infoLog);
		std::cout << "Error linking shaders: " << infoLog << std::endl;
	}
}
Shader::Shader(const char* vertexPath, const char* fragmentPath){
	std::string vertexCode;
	std::string fragmentCode;
	
	std::ifstream vShaderFile;
	std::ifstream fShaderFile;
	
    vShaderFile.exceptions (std::ifstream::failbit | std::ifstream::badbit);
    fShaderFile.exceptions (std::ifstream::failbit | std::ifstream::badbit);
	
	try{
		vShaderFile.open(vertexPath);
		fShaderFile.open(fragmentPath);
		
		std::stringstream vShaderStream, fShaderStream;
		vShaderStream << vShaderFile.rdbuf();
		fShaderStream << fShaderFile.rdbuf();
		
		vShaderFile.close();
		fShaderFile.close();
		
		vertexCode = vShaderStream.str();
		fragmentCode = fShaderStream.str();
	}catch(std::ifstream::failure e){
		std::cout << "Cannot read file!" << std::endl;
	}
	
	const char* vShaderCode = vertexCode.c_str();
	const char* fShaderCode = fragmentCode.c_str();
	
	GLuint vertexShader = glCreateShader(GL_VERTEX_SHADER);
	glShaderSource(vertexShader, 1, &vShaderCode, NULL);
	glCompileShader(vertexShader);

	shaderErrorCheck(&vertexShader);
	
	GLuint fragmentShader = glCreateShader(GL_FRAGMENT_SHADER);
	glShaderSource(fragmentShader, 1, &fShaderCode, NULL);
	glCompileShader(fragmentShader);

	shaderErrorCheck(&fragmentShader);
	
	ID = glCreateProgram();
	glAttachShader(ID, vertexShader);
	glAttachShader(ID, fragmentShader);
	glLinkProgram(ID);
	
	programErrorCheck(&ID);
	
	glDeleteShader(vertexShader);
	glDeleteShader(fragmentShader);
}

void Shader::use(){
	glUseProgram(ID);
}

void Shader::setBool(const std::string &name, bool value) const{
	glUniform1i(glGetUniformLocation(ID, name.c_str()), (int)value);
}
void Shader::setInt(const std::string &name, int value) const{
	glUniform1i(glGetUniformLocation(ID, name.c_str()), value);
}  
void Shader::setFloat(const std::string &name, float value) const{
	glUniform1i(glGetUniformLocation(ID, name.c_str()), value);
}

File: graphics/Shader.java
package graphics;

import static org.lwjgl.opengl.GL33.*;
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

public class Shader {
	
	private final int program = glCreateProgram();
	
	private static void ShaderErrorCheck(int shader) throws Exception {
		int success = glGetShaderi(shader, GL_COMPILE_STATUS);
		if(success == 0){
			System.out.println("Shader error check failed.");
			String infoLog = glGetShaderInfoLog(shader);
			throw new Exception(infoLog);
		}
	}
	
	private static void ProgramErrorCheck(int program) throws Exception {
		int success = glGetProgrami(program, GL_LINK_STATUS);
		if(success == 0){
			System.out.println("Program error check failed.");
			String infoLog = glGetProgramInfoLog(program);
			throw new Exception(infoLog);
		}
	}
	
	private static String ReadSource(String path) {
		String data = "";
		try {
			File source = new File(path);
			Scanner scan = new Scanner(source);
			
			while(scan.hasNextLine()) {
				data = data + scan.nextLine() + "\n";
			}
			scan.close();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
			System.exit(-1);
		}
		return data;
	}
	
	public void Use() {
		glUseProgram(program);
	}
	
	public int GetProgramID() {
		return program;
	}
	
	//The value needs to be passed as either 1 or 0.
	public void SetBool(String name, int value) {
		glUniform1i(glGetUniformLocation(program, name), value);
	}
	
	public void SetInt(String name, int value) {
		glUniform1i(glGetUniformLocation(program, name), value);
	}
	
	public void SetFloat(String name, int value) {
		glUniform1f(glGetUniformLocation(program, name), value);
	}
	
	public Shader(String vertexPath, String fragmentPath) {
		try {
			int vertexShader = glCreateShader(GL_VERTEX_SHADER);
			glShaderSource(vertexShader, ReadSource(vertexPath));
			glCompileShader(vertexShader);
	
			ShaderErrorCheck(vertexShader);
			
			int fragmentShader = glCreateShader(GL_FRAGMENT_SHADER);
			glShaderSource(fragmentShader, ReadSource(fragmentPath));
			glCompileShader(fragmentShader);
	
			ShaderErrorCheck(fragmentShader);
			
			glAttachShader(program, vertexShader);
			glAttachShader(program, fragmentShader);
			glLinkProgram(program);
			
			ProgramErrorCheck(program);
			
			glDeleteShader(vertexShader);
			glDeleteShader(fragmentShader);
		}catch (Exception e) {
			System.out.println(e);
			System.exit(-1);
		}
	}
	
	public void Delete() {
		glDeleteProgram(program);
	}
}
